### PR TITLE
Fix http response input stream resource leak

### LINKED ISSUE #624: Memory leak in HttpResponseStreamHandler
The problem is similar to #560 but happens with `HttpResponseStreamHandler`. It's inner class `HttpResponseInputStream` doesn't seem to manage `ByteBuf current`.

```
ERROR - etty.util.ResourceLeakDetector - LEAK: ByteBuf.release() was not called before it's garbage-collected. See http://netty.io/wiki/reference-counted-objects.html for more information.
WARNING: 1 leak records were discarded because the leak record count is limited to 4. Use system property io.netty.leakDetection.maxRecords to increase the limit.
Recent access records: 4
#4:
    io.netty.buffer.AdvancedLeakAwareByteBuf.readByte(AdvancedLeakAwareByteBuf.java:399)
    com.github.dockerjava.netty.handler.HttpResponseStreamHandler$HttpResponseInputStream.read(HttpResponseStreamHandler.java:84)
    java.io.InputStream.read(InputStream.java:179)
    java.io.InputStream.read(InputStream.java:101)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:67)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:48)
    ...
    java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    java.util.concurrent.FutureTask.run(FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    java.lang.Thread.run(Thread.java:745)
#3:
    io.netty.buffer.AdvancedLeakAwareByteBuf.readByte(AdvancedLeakAwareByteBuf.java:399)
    com.github.dockerjava.netty.handler.HttpResponseStreamHandler$HttpResponseInputStream.read(HttpResponseStreamHandler.java:84)
    java.io.InputStream.read(InputStream.java:170)
    java.io.InputStream.read(InputStream.java:101)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:67)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:48)
    ...
    java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    java.util.concurrent.FutureTask.run(FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    java.lang.Thread.run(Thread.java:745)
#2:
    io.netty.buffer.AdvancedLeakAwareByteBuf.readByte(AdvancedLeakAwareByteBuf.java:399)
    com.github.dockerjava.netty.handler.HttpResponseStreamHandler$HttpResponseInputStream.read(HttpResponseStreamHandler.java:84)
    java.io.InputStream.read(InputStream.java:179)
    java.io.InputStream.read(InputStream.java:101)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:67)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:48)
    ...
    java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    java.util.concurrent.FutureTask.run(FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    java.lang.Thread.run(Thread.java:745)
#1:
    io.netty.buffer.AdvancedLeakAwareByteBuf.readByte(AdvancedLeakAwareByteBuf.java:399)
    com.github.dockerjava.netty.handler.HttpResponseStreamHandler$HttpResponseInputStream.read(HttpResponseStreamHandler.java:84)
    java.io.InputStream.read(InputStream.java:170)
    java.io.InputStream.read(InputStream.java:101)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:67)
    org.apache.commons.compress.utils.IOUtils.copy(IOUtils.java:48)
    ...
    java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    java.util.concurrent.FutureTask.run(FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    java.lang.Thread.run(Thread.java:745)
Created at:
    io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:271)
    io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
    io.netty.buffer.UnsafeByteBufUtil.copy(UnsafeByteBufUtil.java:464)
    io.netty.buffer.PooledUnsafeDirectByteBuf.copy(PooledUnsafeDirectByteBuf.java:309)
    io.netty.buffer.PooledSlicedByteBuf.copy(PooledSlicedByteBuf.java:96)
    io.netty.buffer.AbstractByteBuf.copy(AbstractByteBuf.java:1161)
    com.github.dockerjava.netty.handler.HttpResponseStreamHandler.channelRead0(HttpResponseStreamHandler.java:30)
    com.github.dockerjava.netty.handler.HttpResponseStreamHandler.channelRead0(HttpResponseStreamHandler.java:20)
    io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:334)
    io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:326)
    com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:72)
    com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:33)
    io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:334)
    io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:326)
    io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:233)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:334)
    io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:326)
    io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:428)
    io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
    io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
    io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:243)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:334)
    io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:326)
    io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1320)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:334)
    io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:905)
    io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:957)
    io.netty.channel.epoll.EpollDomainSocketChannel$EpollDomainUnsafe.epollInReady(EpollDomainSocketChannel.java:140)
    io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:384)
    io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:300)
    io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
    io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145)
    java.lang.Thread.run(Thread.java:745) 
```

### PR DESCRIPTION
**Pull request by tejksat**
According to https://github.com/netty/netty/wiki/Reference-counted-objects tests logs should be checked for an error message about a memory leak:
> PARANOID - Same with ADVANCED except that it's for every single buffer. Useful for automated testing phase. You could fail the build if the build output contains 'LEAK:'.

I don't know how to implement it in a good way. Could you please suggest?

Fixes #624

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/633)

### CODE REVIEW

**Comment by tejksat** [comment]
@marcuslinke @KostyaSha could you please check the changes? Please also note the question in the comment above.

**Comment by KostyaSha** [comment]
Sorry, lost comment. Could you check coverity results for master branch? Does it has complains about this resource leak?  https://scan.coverity.com/projects/docker-java-docker-java?tab=overview

**Comment by tejksat** [comment]
@KostyaSha sure! I've requested the access to coverty.

**Comment by KostyaSha** [comment]
Hm.. you should be able to see any defects without contributor... Press "view defects"

**Comment by tejksat** [comment]
@KostyaSha when I signed in as GitHub user Coverity shows a message _Want to view defects or help fix defects?_ with a button below _Add me to project_.
image

**Comment by KostyaSha** [comment]
Added as defect viewer, if you will need "triage?" please ping me.

**Comment by KostyaSha** [comment]
@marcuslinke for review

**Comment by tejksat** [comment]
@KostyaSha `HttpResponseStreamHandler` is clean. Does Coverity could find problems with not released Netty buffers?

**Comment by KostyaSha** [comment]
As i understand it analyses only our code and nobody knows what checks it has because it proprietary solution.

**Comment by tejksat** [comment]
Ok.. I'll try to find a solution to check generated logs for messages about `ByteBuf` leaks. I've checked them locally and on downloading a somewhat big file there is a resource leak message there.

**Comment by tejksat** [comment]
@KostyaSha @marcuslinke could you please check? Travis errors don't seem to be related with PR changes.

**Comment by KostyaSha** [comment]
@tejksat 1.11 tcp build is green and it fine, 1.12 tcp shouldn't have new failures (atm only #670 fails)

File: src/main/java/com/github/dockerjava/core/async/AsyncResultCallbackTemplate.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        } finally {
+            resultReady.countDown();
+        }
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        if (closed) return;
+
+        if (this.firstError == null) {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/async/AsyncResultCallbackTemplate.java]: please don't use `this` where it not required.

File: src/main/java/com/github/dockerjava/netty/handler/HttpResponseStreamHandler.java [hunk 2]
```diff
@@ -90,11 +113,19 @@ public int read() throws IOException {
         private void poll() {
             if (readableBytes() == 0) {
                 try {
-                    current = queue.poll(50, TimeUnit.MILLISECONDS);
+                    releaseCurrent();
+                    current = queue.poll(500, TimeUnit.MILLISECONDS);
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/netty/handler/HttpResponseStreamHandler.java]: 50 to 500?

Comment (tejksat) [inline src/main/java/com/github/dockerjava/netty/handler/HttpResponseStreamHandler.java]: There were problems on getting the first data chunk from the queue. I'm looking forward to replace `queue` with `Exchanger`. This code would gone.

**Comment by tejksat** [comment]
@KostyaSha pls review!

**Comment by tejksat** [comment]
@KostyaSha pls check! Hope this is the final version :)

**Comment by KostyaSha** [comment]
@tejksat https://github.com/codecov/browser-extension if you would like to cover all your logic :)

**Comment by KostyaSha** [comment]
@tejksat i don't understand this routines, would wait bit for @marcuslinke or merge as it passes tests.

**Comment by tejksat** [comment]
@KostyaSha codecov Chrome extension is the best! Thank you!

**Comment by marcuslinke** [comment]
@tejksat Sorry for being late but I don't get the intent of this PR. As I understood from your intitial comment on #624 the problem was that the `current` `ByteBuf` wasn't released properly. This could be easily fixed by calling `current.discardReadBytes()`. Could you explain please?

**Comment by tejksat** [comment]
@marcuslinke there are several problems in the original code:
- not releasing `ByteBuf` queued;
- queuing `ByteBuf` until all data is read what effectively leads to `OutOfDirectMemoryError` if the data read is too big. This happens if you try to copy too big file or folder from a container. To fix this `ResponseCallback` (which in `awaitResult()` waits in fact for `readComplete()` channel method invocation) replaced with `AsyncResultCallback`;
- actually queuing `ByteBuf` anyway leads to `OutOfDirectMemoryError` if data consumption is slower (the next point) than obtaining it;
- poor performance because of the byte-by-byte reads.

File: src/main/java/com/github/dockerjava/core/async/AsyncResultCallback.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+
+import com.github.dockerjava.api.async.ResultCallback;
+import com.google.common.base.Throwables;
+
+/**
+ * Implementation of {@link ResultCallback} with the single result event expected.
+ */
+public class AsyncResultCallback<A_RES_T> implements ResultCallback<A_RES_T> {
```
Comment (marcuslinke) [inline src/main/java/com/github/dockerjava/core/async/AsyncResultCallback.java]: This special callback class should inherit `ResultCallbackTemplate` ideally and should be moved into `InvocationBuilder` internally.

File: src/main/java/com/github/dockerjava/netty/handler/HttpResponseStreamHandler.java [hunk 4]
```diff
@@ -19,45 +16,87 @@
...
-            queue.put(byteBuf);
+        private final Object lock = new Object();
+
+        public void write(ByteBuf byteBuf) throws InterruptedException {
+            synchronized (lock) {
+                if (closed) {
+                    return;
+                }
+                while (current != null) {
+                    lock.wait();
```
Comment (marcuslinke) [inline src/main/java/com/github/dockerjava/netty/handler/HttpResponseStreamHandler.java]: This blocks the call to `write` until the `current` buffer is read completely, right?

Comment (tejksat) [inline src/main/java/com/github/dockerjava/netty/handler/HttpResponseStreamHandler.java]: Yes, that's right. Or until this `HttpResponseInputStream` is closed.

File: src/main/java/com/github/dockerjava/core/async/AsyncResultCallback.java (hunk 3 above)
Comment (tejksat) [inline src/main/java/com/github/dockerjava/core/async/AsyncResultCallback.java]: Done!

**Comment by marcuslinke** [comment]
LGTM. Thanks @tejksat for fixing all these issues!

**Comment by tejksat** [comment]
@marcuslinke I'm glad to do that!

**Pull request merged**